File: DYReactImplTest/android/app/src/main/java/com/dyreactimpltest/bridge/DYReact.java
package com.dyreactimpltest.bridge;

import android.telecom.Call;

import androidx.annotation.NonNull;

import com.dynamicyield.dyapi.DYApi;
import com.dynamicyield.dyapi.DYApiResults.DYEvaluatorSet;
import com.dynamicyield.dyapi.DYApiResults.DYSetSecretResult;
import com.dynamicyield.dyapi.DYApiResults.DYTrackResult;
import com.dynamicyield.engine.DYCustomSettings;
import com.dynamicyield.engine.DYPageContext;
import com.dynamicyield.engine.DYRecommendationListenerItf;
import com.dynamicyield.state.DYExperimentsState;
import com.dynamicyield.userdata.external.DYUserData;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;


public class DYReact extends ReactContextBaseJavaModule{
    private static ReactApplicationContext mReactContext;

    private boolean setSecretCallback = false;

    DYReact(ReactApplicationContext context) {
        super(context);
        mReactContext = context;
    }

    @NonNull
    @Override
    public String getName() {
        return "DYReact";
    }

    @ReactMethod
    public void setSecret(String secretKey,Callback callback) {
        DYApi.setContextAndSecret(mReactContext, secretKey, new DYSetSecretResult() {
            @Override
            public void onEnd(DYExperimentsState dyExperimentsState) {
                if (!setSecretCallback && dyExperimentsState.ordinal() >= 2) {
                    setSecretCallback = true;
                    callback.invoke(dyExperimentsState.ordinal());
                }
            }
        });
    }

    @ReactMethod
    public void pageView(String uniqueID, ReadableMap context, Callback callback) {
        DYPageContext dyPageContext = convertReadableMapToDYContext(context,callback);

        DYApi.getInstance().trackPageView(uniqueID, dyPageContext, new DYTrackResult() {
            @Override
            public void onEnd(String pageName) {
                callback.invoke(pageName);
            }
        });
    }

    @ReactMethod
    public void trackEvent(String eventName, ReadableMap params, Callback callback) {
        JSONObject jsonParams = (params != null) ? new JSONObject(params.toHashMap()) : null;

        DYApi.getInstance().trackEvent(eventName,jsonParams, new DYTrackResult() {
            @Override
            public void onEnd(String pageName) {
                callback.invoke(pageName);
            }
        });
    }

    @ReactMethod
    public void identifyUser(ReadableMap userData, Callback callback){
        if (userData != null) {
            DYUserData dyUserData = new DYUserData(new JSONObject(userData.toHashMap()));
            DYApi.getInstance().identifyUser(dyUserData);
        } else {
            callback.invoke("No user data supplied");
        }
    }

    @ReactMethod
    public void consentOptOut() {
        DYApi.getInstance().consentOptOut();
    }

    @ReactMethod
    public void consentOptIn() {
        DYApi.getInstance().consentOptIn();
    }

    @ReactMethod
    public void setEvaluator(String evaluatorID, ReadableArray evaluatorData, boolean saveBetweenSessions, Callback callback){
        JSONArray data = evaluatorData != null ? new JSONArray(evaluatorData.toArrayList()) : null;

        DYApi.getInstance().setEvaluator(evaluatorID, data, saveBetweenSessions, new DYEvaluatorSet() {
            @Override
            public void onEvaluatorSet(String evaluatorID, JSONArray evaluatorData, boolean saveBetweenSessions) {
                try {
                    callback.invoke(evaluatorID, convertJsonToArray(evaluatorData), saveBetweenSessions);
                } catch (JSONException e) {

                }
            }
        });
    }

    @ReactMethod
    public void getSmartVariable(String varName, String defaultValue, Callback callback){
        Object result = DYApi.getInstance().getSmartVariable(varName, defaultValue);

        callback.invoke(varName,result instanceof String ? (String)result : (Double)result);
    }

    @ReactMethod
    public void sendRecommendationRequest(String widgetID,ReadableMap context, Callback callback){
        DYPageContext dyPageContext = convertReadableMapToDYContext(context,callback);

        DYApi.getInstance().sendRecommendationsRequest(widgetID, dyPageContext, new DYRecommendationListenerItf() {
            @Override
            public void onRecommendationResult(JSONArray jsonArray, String widgetID) {
                try {
                    callback.invoke(convertJsonToArray(jsonArray),widgetID);
                } catch (JSONException e) {
                    callback.invoke(null,widgetID);
                }
            }
        });
    }

    @ReactMethod
    public void trackRecomItemClick(String widgetID, String itemId) {
        DYApi.getInstance().trackRecomItemClick(widgetID,itemId);
    }

    @ReactMethod
    public void trackRecomItemRealImpression(String widgetID, ReadableArray itemIds) {
        DYApi.getInstance().trackRecomItemRealImpression(widgetID,itemIds.toArrayList().toArray(new String[0]));
    }

    @ReactMethod
    public void setUseEuropeanServer(boolean on){
        DYCustomSettings settings = new DYCustomSettings();
        settings.useEuropeanServer(on);
        DYApi.setCustomSettings(settings);
    }

    private DYPageContext convertReadableMapToDYContext(ReadableMap context, Callback callback){
        DYPageContext dyPageContext = null;
        if (context != null) {
            dyPageContext = new DYPageContext(new JSONObject(context.toHashMap()));
        } else {
            callback.invoke("ERROR");
        }


        return dyPageContext;
    }

    private static WritableMap convertJsonToMap(JSONObject jsonObject) throws JSONException {
        WritableMap map = new WritableNativeMap();

        Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            String key = iterator.next();
            Object value = jsonObject.get(key);
            if (value instanceof JSONObject) {
                map.putMap(key, convertJsonToMap((JSONObject) value));
            } else if (value instanceof  JSONArray) {
                map.putArray(key, convertJsonToArray((JSONArray) value));
            } else if (value instanceof  Boolean) {
                map.putBoolean(key, (Boolean) value);
            } else if (value instanceof  Integer) {
                map.putInt(key, (Integer) value);
            } else if (value instanceof  Double) {
                map.putDouble(key, (Double) value);
            } else if (value instanceof String)  {
                map.putString(key, (String) value);
            } else {
                map.putString(key, value.toString());
            }
        }
        return map;
    }

    private static WritableArray convertJsonToArray(JSONArray jsonArray) throws JSONException {
        WritableArray array = new WritableNativeArray();

        for (int i = 0; i < jsonArray.length(); i++) {
            Object value = jsonArray.get(i);
            if (value instanceof JSONObject) {
                array.pushMap(convertJsonToMap((JSONObject) value));
            } else if (value instanceof  JSONArray) {
                array.pushArray(convertJsonToArray((JSONArray) value));
            } else if (value instanceof  Boolean) {
                array.pushBoolean((Boolean) value);
            } else if (value instanceof  Integer) {
                array.pushInt((Integer) value);
            } else if (value instanceof  Double) {
                array.pushDouble((Double) value);
            } else if (value instanceof String)  {
                array.pushString((String) value);
            } else {
                array.pushString(value.toString());
            }
        }
        return array;
    }
}

File: DYReactImplTest/ios/DYReact.m
//
//  DYReact.m
//  DYReactImplTest
//
//  Created by Idan Oshri on 23/12/2019.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "DYReact.h"

@implementation DYReact

RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(setSecret:(NSString*)secret callback:(RCTResponseSenderBlock)callback){
  [[DYApi getInstance] setSecretKey:secret completionHandler:^(ExperimentsState state) {
    if (state == DY_READY_AND_UPDATED || state == DY_READY_NO_UPDATE_NEEDED){
      static dispatch_once_t onceToken;
      dispatch_once(&onceToken, ^{
        callback(@[[NSString stringWithFormat:@"%d",state]]);
      });
    }
  }];
}

RCT_EXPORT_METHOD(trackEvent:(NSString*)eventName params:(NSDictionary*)params callback:(RCTResponseSenderBlock)callback){
  [[DYApi getInstance] trackEvent:eventName prop:params completionHandler:^(NSString * _Nonnull name) {
    callback(@[name]);
  }];
}

RCT_EXPORT_METHOD(pageView:(NSString* _Nonnull)uniqueID context:(NSDictionary* _Nullable)context callback:(RCTResponseSenderBlock)callback){
  DYPageContext* dyContext = [[DYPageContext alloc] initWithJSONDic:context];
  [[DYApi getInstance] pageView:uniqueID context:dyContext completionHandler:^(NSString * _Nonnull name) {
    callback(@[name]);
  }];
}

RCT_EXPORT_METHOD(consentOptIn){
  [[DYApi getInstance] consentOptin];
}

RCT_EXPORT_METHOD(consentOptOut){
  [[DYApi getInstance] consentOptout];
}

RCT_EXPORT_METHOD(setEvaluator:(NSString* _Nonnull)evaluatorID forParams:(NSArray* _Nonnull)params saveBetweenSessions:(BOOL)save  callback:(RCTResponseSenderBlock)callback){
  
  [[DYApi getInstance] setEvaluator:evaluatorID forParams:params saveBetweenSessions:save completionHandler:^(NSString * _Nonnull evaluatorID, NSArray * _Nonnull params, BOOL isPersistent) {
    callback(@[evaluatorID,params,[NSNumber numberWithBool:isPersistent]]);
  }];
}

RCT_EXPORT_METHOD(getSmartObjectData:(id _Nonnull)smartObjectID callback:(RCTResponseSenderBlock)callback){
  id result = [[DYApi getInstance] getSmartObjectData:smartObjectID];
  callback(@[smartObjectID,result]);
}

RCT_EXPORT_METHOD(getSmartVariable:(NSString* _Nonnull)varName defaultValue:(id _Nullable)defaultValue callback:(RCTResponseSenderBlock)callback){
  id result = [[DYApi getInstance] getSmartVariable:varName defaultValue:defaultValue];
  callback(@[varName,result]);
}

RCT_EXPORT_METHOD(identifyUser:(NSDictionary*)userData callback:(RCTResponseSenderBlock)callback){
  if (userData){
    DYUserData* dyUserData = [DYUserData createFromDictionary:userData];
    [[DYApi getInstance] identifyUser:dyUserData];
  } else {
    callback(@[@"No user data supplied"]);
  }
}

RCT_EXPORT_METHOD(sendRecommendationRequest:(NSString* _Nonnull)widgetID context:(NSDictionary* _Nullable)context callback:(RCTResponseSenderBlock)callback){
  DYPageContext* dyContext = [[DYPageContext alloc] initWithJSONDic:context];
  [[DYApi getInstance] sendRecommendationRequest:widgetID withContext:dyContext completionHandler:^(NSArray * _Nullable recommendations, NSString * _Nonnull widgetID) {
    callback(@[recommendations,widgetID]);
  }];
}

RCT_EXPORT_METHOD(trackRecomItemRealImpression:(NSString* _Nonnull)widgetID itemIDs:(NSArray* _Nonnull)itemIDs){
  [[DYApi getInstance] TrackRecomItemsRealImpression:widgetID andItemID:itemIDs];
}

RCT_EXPORT_METHOD(trackRecomItemClick:(NSString* _Nonnull)widgetID itemID:(NSString* _Nonnull)itemID){
  [[DYApi getInstance] TrackRecomItemClick:widgetID andItemID:itemID];
}

RCT_EXPORT_METHOD(setUseEuropeanServer:(BOOL)on){
  [[DYApi getInstance] setUseEuropeanServer:on];
}

@end
